datafusion/physical-plan/src/memory.rs

```diff
@@ -365,8 +366,165 @@ impl RecordBatchStream for MemoryStream {
     }
 }
 
+pub trait StreamingBatchGenerator: Send + Sync + fmt::Debug + fmt::Display {
```

Comment:
I am not very satisfied with these `Streaming... `naming. Can we describe the behavior better with a different naming for these? Perhaps we could use 'lazy' term